packages/webapp/src/containers/Animals/AddAnimals/AddAnimalDetails/index.tsx

```diff
@@ -13,8 +13,221 @@
  *  GNU General Public License for more details, see <https://www.gnu.org/licenses/>.
  */
 
+import { ReactElement } from 'react';
+import { useFieldArray, useFormContext } from 'react-hook-form';
+import { useTranslation } from 'react-i18next';
+import styles from './styles.module.scss';
+import { Info, Semibold } from '../../../../components/Typography';
+import ExpandableItem from '../../../../components/Expandable/ExpandableItem';
+import useExpandable from '../../../../components/Expandable/useExpandableItem';
+import { AnimalFormHeaderItem } from '../../../../components/Animals/AddAnimalsForm/AnimalFormHeaderItem';
+import AnimalDetails from '../../../../components/Animals/AddAnimalsDetails';
+import BatchDetails from '../../../../components/Animals/AddBatchDetails';
+import { useCurrencySymbol } from '../../../hooks/useCurrencySymbol';
+import { useAnimalOptions } from '../useAnimalOptions';
+import { STEPS } from '..';
+import { DetailsFields, Option } from '../types';
+import { AddAnimalsFormFields } from '../types';
+import { AnimalOrBatchKeys, AnimalOrigins } from '../../types';
+import { parseUniqueDefaultId } from '../../../../util/animal';
+import { getDefaultAnimalIconName } from '../../Inventory/useAnimalInventory';
+import usePopulateDetails from './usePopulateDetails';
+
 const AddAnimalDetails = () => {
-  return <div>Add animal details</div>;
+  const { expandedIds, toggleExpanded } = useExpandable({ isSingleExpandable: true });
+  const { t } = useTranslation();
+  const { control, getValues, watch } = useFormContext<AddAnimalsFormFields>();
+
+  const { fields, remove, replace } = useFieldArray({
+    name: STEPS.DETAILS,
+    control,
+  });
+
+  const onRemoveCard = (index: number): void => {
+    remove(index);
+  };
+
+  const {
+    defaultTypes,
+    sexOptions,
+    sexDetailsOptions,
+    useOptions,
+    tagTypeOptions,
+    tagColorOptions,
+    organicStatusOptions,
+    originOptions,
+  } = useAnimalOptions(
+    'default_types',
+    'sex',
+    'sexDetails',
+    'use',
+    'tagType',
+    'tagColor',
+    'organicStatus',
+    'origin',
+  );
+
+  /* Populate details form based on basics data */
+  usePopulateDetails(getValues, replace);
+
+  /* Render logic */
+  const animalElements: ReactElement[] = [];
+  const batchElements: ReactElement[] = [];
+
+  const generalDetailProps = {
+    sexOptions,
+    sexDetailsOptions,
+  };
+
+  const otherDetailsProps = {
+    organicStatusOptions,
+  };
+
+  const currency = useCurrencySymbol();
+  const originProps = {
+    currency: currency,
+    originOptions,
+  };
+
+  let animalIndex = 1;
+  let batchIndex = 1;
+
+  const animalCount = fields.filter(
+    (field) => field.animal_or_batch === AnimalOrBatchKeys.ANIMAL,
+  ).length;
+  const batchCount = fields.filter(
+    (field) => field.animal_or_batch === AnimalOrBatchKeys.BATCH,
+  ).length;
+
+  fields.forEach((field, index) => {
```

Comment:
😭 It was such a bad idea to have the container hold the logic!!! I think I wasn't sure what the API responses for uses, sexes or origin would be like, so I went with an approach that wouldn't be affected by the API response structure 😞 Is it an option to move the components? 😢 

I didn't test it, but my approach for this component would be:
1. Create a function to generate `ExpandableItem` depending on `animal_or_batch` (almost like the callback function of the forEach).
2. Separate fields into animals and batches.
3. Render animals and batches like this?:
```
{animals.map((animal, index) => generateExpandableItem(animal, index))}
{batches.map((batch, index) => generateExpandableItem(batch, index))}
```

I think it would be easier to handle `animalIndex` and `batchIndex`.

Apart from that, I think `field.animal_or_batch === AnimalOrBatchKeys.ANIMAL` can be replaced with `isAnimal`  :)

Reply:
I will most happily move the logic into the components! 😄🙏 

Thank you for the render suggestion. That manual index increment felt like such a code smell but it was a long function and hard for me to think about. I'll take a fresh look after the other code cleanup is completed and try to implement what you describe above ❤️

Reply:
@SayakaOno I did exactly what you said (made the callback of the `forEach` a separate function and mapped over in the render itself) and I already think it's SOOO much better!! But could you take a look to see if it's implemented correctly according to what you were thinking? 🙏

Reply:
Thank you for moving the logic, it's much cleaner 😭 
`fieldArrayIndex` is something that I didn't take into account, thank you for handling it! Other than that, the new code is exactly what I was picturing. Thank you!!